### ISSUE TITLE
Support for custom object mappers with Jackson?

### ISSUE THREAD

**Issue by vincent-zurczak** [stoicflame/enunciate#314]
Hi! I find Enunciate very convenient as it generates quite a lot of stuff without any intrusive annotations. However, I am working on a project that has a more complex structure than a simple microservice. The model used by the REST API was (and should not) be annotated. It is defined in its own Maven module and is used by 5 or 6 projects. The JSon generation is tightly controlled in a separate Maven module through custom object mappers. You can find an example of such a mapper [here](https://github.com/roboconf/roboconf-platform/blob/606223dbaae4638facd7ad6605a222aa972e5518/core/roboconf-dm-rest-commons/src/main/java/net/roboconf/dm/rest/commons/json/JSonBindingUtils.java).

I was wondering whether there was a solution to generate the right JSon models from such a class. The default generation includes fields that are concretely not visible by the REST API. In my **enunciate.xml** file, the JAX-B module is disabled and the jackson1 module does not honor JAX-B either.

If it is not possible, is there a way to manually specify the JSon shape?
This could be something raw like the exact JSon model.

```properties
my.package.MyClass = { "field1": "value1", "field2" : "value2"...  }
```
Or this could be something that could be used by the Jackson module. E.g a Java class defined in the current project, or a properties file. Something like...

```java
ObjectMapper getCustomObjectMapper() {
          return new MyObjectMapper();
}

Map<Class<?>,Object> getCustomTypeSamples() {

          Map<Class<?>,Object> result = new HashMap<> ();
          result.put( MyClass1.class, new MyClass1( sample parameters );
          ...
          
          return result;
}
```

Roughly, the Jackson module would use the map and the custom object mapper to "deduce" the shape of the JSonified model to inject it in the documentation. When nothing is specified for a class, we could use the usual module resolution.

I have not found anything for such a case, but I might have missed it.
Thanks for your help.

**Comment by stoicflame** [comment]
There's nothing currently available, but we'll keep this issue open as an enhancement request.

**Comment by vincent-zurczak** [comment]
Thanks. :+1:

**Comment by dklotz** [comment]
We're currently using some custom Jackson modules to de/serialize some more advanced objects to/from JSON (e.g. adding some custom fields during serialization, using these fields to map to a specific DTO sub-class during deserialization etc.), so some easy way to just specify "include these custom jackson module(s)" (or a way to provide a custom factory for the Jackson ObjectMapper) would be great.

**Comment by hiddendog** [comment]
Is there at least a way to tell Enunciate to understand Jackson's mixins?

**Comment by stoicflame** [comment]
> Is there at least a way to tell Enunciate to understand Jackson's mixins? 

There's not yet, no. See #472.

**Comment by hiddendog** [comment]
Maybe I can help, any idea where I should start, which classes I should be looking at, etc?

**Comment by stoicflame** [comment]
I'll try to provide some pointers in the right direction. First, note that the following is about the "jackson" module, but the same needs to be applied to the "jackson1" module as well (both places).

The adjustment will be made in the `com.webcohesion.enunciate.modules.jackson.model.types.JsonTypeFactory#findSpecifiedType` method, which will need to look in the config file (available via the `EnunciateJacksonContext`) to see what mix-ins are configured. Basically, if a mix-in is configured, say that the specified type is the mix-in and not the original class.

Hope that helps. Thanks!

**Comment by vincent-zurczak** [comment]
I have spent most of the day thinking and working on it.
And I do not think there is any reasonable solution for this problem. Object mappers can produce direct JSon as text. But Enunciate uses a recursive structure (JsonType, and in particular DataType as a sub-class). It is not possible to map a direct string into this structure.

I also tried to use Swagger Core directly, outside Enunciate.
I was thinking I could perform a 2-pass processing, one with Enunciate, and a second one with Swagger Core to update the swagger.json file. That will not work either. Just like Enunciate, Swagger Core uses a (very similar) recursive structure for type definitions.

IMO, this issue cannot be solved.
What we should expect now is a workaround. #472 should allow to use Jackson mixins. They are annotations. It would be nice if we could also parse them as doclets. Using `@JsonIgnore` and 
`@JsonProperty` as doclets, we would have a non-intrusive workaround to customize the definitions generated by Enunciate. Of course, it would somehow duplicate part of the job made in the object mapper... but it is better than nothing. :wink: 

As a reminder, we chose object mappers because our model is used but not focused on a REST API.
So, we cannot add Jackson annotations in the model. It is defined in its own Maven module. The binding with Jackson is made in a different project. Doclets would allow to support such a use case.

If you agree with that, I suggest to complete #472 and close this issue with the **won't fix** label.

**Comment by stoicflame** [comment]
Agreed.

Added support for mixins at 7052d14:

```xml
<enunciate>
  <modules>
    <jackson>
      <mixin target="com.thirdparty.ExternalClass" source="com.mycompany.ExternalClassMixin"/>
```

Also works for `jackson1` module.

**Comment by vincent-zurczak** [comment]
Hi! I have experimented the use of Jackson mix-ins for this use case.
And it works well, that's cool.

I however found a strong limitation to mix-ins. This limit does not only apply to Enunciate, but also to Jackson itself. Mixins are good only when one wants to remove or rename JSon properties. But they cannot be used to add new properties. Said differently, they do not exactly replace custom object mappers, which remain the last resort solution.

I finally resigned myself for a two-step approach.

1. Generate the swagger.json file with Enunciate's Maven plug-in.
2. Load and update the generated file in the Maven's build process thanks to ANT.

Since custom object mappers are somehow workaround themselves, it is not surprising we may have to code some little things for Swagger. The main idea is that we do not create type definitions by type examples. The generated JSon file is loaded without any error in Swagger UI.

I wrote an article on [my blog](https://vzurczak.wordpress.com/2016/08/24/generating-swagger-json-files-with-enunciate-and-custom-objects-mappers/) about it, as it would be too long to put the full solution here. I think these cases are unusual and over there is a solution for them.
